persistence/src/main/java/de/cyface/persistence/MeasurementTable.java

```diff
@@ -96,57 +112,154 @@ public void onUpgrade(final SQLiteDatabase database, final int fromVersion, fina
+        try {
+            measurementCursor = database.query("measurements", new String[] {"_id"}, null, null, null, null, null,
+                    null);
+            if (measurementCursor.getCount() == 0) {
+                Log.v(TAG, "No measurements for migration found");
+                return;
+            }
+
+            // Check all measurements
+            while (measurementCursor.moveToNext()) {
+                final int identifierColumnIndex = measurementCursor.getColumnIndex("_id");
+                final long measurementId = measurementCursor.getLong(identifierColumnIndex);
+
+                geoLocationCursor = database.query("locations",
+                        new String[] {"lat", "lon", "gps_time", "speed", "accuracy"}, "measurement_fk = ?",
+                        new String[] {String.valueOf(measurementId)}, null, null, "gps_time ASC", null);
+                if (geoLocationCursor.getCount() < 2) {
+                    Log.v(TAG, "Not enough geoLocations to update distance in measurement entry:" + measurementId);
+                    continue;
+                }
+
+                // Calculate distance for selected measurement
+                final DistanceCalculationStrategy distanceCalculationStrategy = new DefaultDistanceCalculationStrategy();
+                double distance = 0.0;
+                GeoLocation previousLocation = null;
+                while (geoLocationCursor.moveToNext()) {
+                    final int latColumnIndex = geoLocationCursor.getColumnIndex("lat");
+                    final int lonColumnIndex = geoLocationCursor.getColumnIndex("lon");
+                    final int timeColumnIndex = geoLocationCursor.getColumnIndex("gps_time");
+                    final int speedColumnIndex = geoLocationCursor.getColumnIndex("speed");
+                    final int accuracyColumnIndex = geoLocationCursor.getColumnIndex("accuracy");
+                    final double lat = geoLocationCursor.getFloat(latColumnIndex);
+                    final double lon = geoLocationCursor.getFloat(lonColumnIndex);
+                    final long time = geoLocationCursor.getLong(timeColumnIndex);
+                    final float speed = geoLocationCursor.getFloat(speedColumnIndex);
+                    final int accuracy = geoLocationCursor.getInt(accuracyColumnIndex);
+                    final GeoLocation geoLocation = new GeoLocation(lat, lon, time, speed, accuracy);
+
+                    // We cannot calculate a distance from just one geoLocation:
+                    if (previousLocation == null) {
+                        previousLocation = geoLocation;
+                        continue;
+                    }
+
+                    // Calculate distance between last two locations
+                    final double newDistance = distanceCalculationStrategy.calculateDistance(previousLocation,
+                            geoLocation);
+                    Validate.isTrue(newDistance >= 0);
```

Comment:
Since there are devices copying the last location multiple times (I guess some Samsung devices) this might fail on those.